Type inference issue for method arguments

Please consider the following code snippet.

```java
public class Base{
	class Derived extends Base{
	}
	
	public void bar(Optional<Base> o) {
	}

	public void foo() {
		bar(Optional.of(new Derived()));
	}
}
```

For this code the JSS is unable to find a declaration for the method call to `bar` because the argument type is solved as `Optional<Derived>` where Java actually would infer `Optional<Base>`. At least I think that this causes the issue.

+1

I'm seeing similar issues related to type inference, but in a slightly different use-case:

```java

class ComplexGenericArgument {

    static class Items<T> {
        public <Output extends T> Output apply(Transform<? super Items<T>, Output> t) {
            return null;
        }
    }

    static class Transforms {
        public static <I, O> Transform<I, O> of(DoFn<I, O> fn) {
            return null;
        }
    }

    static class Transform<Input, Output> {

    }

    static class DoFn<Input, Output> {

    }

    static class MyFn extends DoFn<Integer, Long> {

    }

    public void main() {
        Items<Integer> items;
        items.apply(Transforms.of(new MyFn())); // this fails to be resolved.
    }

}

```

```
java.lang.RuntimeException: Unable to calculate the type of a parameter of a method call. Method call: items.apply(Transforms.of(new MyFn())), Parameter: Transforms.of(new MyFn())
...
Caused by: java.lang.IndexOutOfBoundsException: Index: 0
	at java.util.Collections$EmptyList.get(Collections.java:4454)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.matchTypeParameters(MethodCallExprContext.java:404)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.resolveMethodTypeParameters(MethodCallExprContext.java:364)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethodAsUsage(MethodCallExprContext.java:92)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:504)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:274)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:39)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:86)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:394)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:286)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:240)
```

I believe I have managed to pinpoint the problem and will have a PR shortly. 
I'm afraid I can't tell for sure whether we are being bitten by the same issue.

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/MethodResolutionLogic.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/MethodResolutionLogic.java
@@ -263,8 +263,9 @@ public class MethodResolutionLogic {
             if (expectedParam.isReferenceType() && actualParam.isReferenceType()) {
                 ResolvedReferenceType r1 = expectedParam.asReferenceType();
                 ResolvedReferenceType r2 = actualParam.asReferenceType();
-
-                return isAssignableMatchTypeParametersMatchingQName(r1, r2, matchedParameters);
+                // we can have r1=A and r2=A.B (with B extends A and B is an inner class of A)
+                // in this case we want to verify expected parameter from the actual parameter ancestors 
+                return isAssignableMatchTypeParameters(r1, r2, matchedParameters);
             }
 
             if (expectedParam.isTypeVariable()) {

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue2062Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue2062Test.java
@@ -1,0 +1,42 @@
+package com.github.javaparser.symbolsolver;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.expr.MethodCallExpr;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+
+public class Issue2062Test extends AbstractSymbolResolutionTest {
+
+    @Test
+    public void test() {
+
+        ParserConfiguration config = new ParserConfiguration();
+        config.setSymbolResolver(new JavaSymbolSolver(new ReflectionTypeSolver(false)));
+        StaticJavaParser.setConfiguration(config);
+
+        String s = "import java.util.Optional;\n" +
+                "public class Base{\n" +
+                "    class Derived extends Base{\n" +
+                "    }\n" +
+                "    \n" +
+                "    public void bar(Optional<Base> o) {\n" +
+                "    }\n" +
+                "    public void foo() {\n" +
+                "        bar(Optional.of(new Derived()));\n" +
+                "    }\n" +
+                "}";
+        CompilationUnit cu = StaticJavaParser.parse(s);
+        List<MethodCallExpr> mces = cu.findAll(MethodCallExpr.class);
+        assertEquals("bar(Optional.of(new Derived()))", mces.get(0).toString());
+        assertEquals("Base.bar(java.util.Optional<Base>)", mces.get(0).resolve().getQualifiedSignature());
+
+    }
+
+}

